File: src/tools/list-workflows.ts
import { getTemporalClient } from "../temporal-client.js";
import type { ListWorkflowsInput, ListWorkflowsOutput, WorkflowSummary } from "../types.js";

function escapeQueryValue(value: string): string {
  return value.replace(/'/g, "''");
}

const STATUS_MAP: Record<string, string> = {
  running: "Running",
  completed: "Completed",
  failed: "Failed",
  timed_out: "TimedOut",
  cancelled: "Canceled",
  terminated: "Terminated",
};

/**
 * Build a Temporal visibility query from the structured input parameters.
 */
function buildQuery(input: ListWorkflowsInput): string {
  // If user provided a raw query, use it directly
  if (input.query) return input.query;

  const clauses: string[] = [];

  if (input.workflow_type) {
    clauses.push(`WorkflowType = '${escapeQueryValue(input.workflow_type)}'`);
  }

  if (input.workflow_id) {
    clauses.push(`WorkflowId = '${escapeQueryValue(input.workflow_id)}'`);
  }

  if (input.status && input.status !== "all") {
    clauses.push(`ExecutionStatus = '${STATUS_MAP[input.status] || input.status}'`);
  }

  if (input.start_time_from) {
    clauses.push(`StartTime >= '${input.start_time_from}'`);
  }

  if (input.start_time_to) {
    clauses.push(`StartTime <= '${input.start_time_to}'`);
  }

  return clauses.join(" AND ");
}

export async function listWorkflows(
  input: ListWorkflowsInput
): Promise<ListWorkflowsOutput> {
  const client = await getTemporalClient(input.namespace);
  const query = buildQuery(input);

  const workflows: WorkflowSummary[] = [];
  let totalCount = 0;
  let hasMore = false;

  try {
    const limit = input.limit || 10;
    let count = 0;

    for await (const workflow of client.workflow.list({
      query: query || undefined,
    })) {
      if (count >= limit) {
        hasMore = true;
        break;
      }

      const startTime = workflow.startTime
        ? workflow.startTime.toISOString()
        : "";
      const closeTime = workflow.closeTime
        ? workflow.closeTime.toISOString()
        : null;
      const durationMs =
        workflow.startTime && workflow.closeTime
          ? workflow.closeTime.getTime() - workflow.startTime.getTime()
          : null;

      // Extract search attributes
      const searchAttributes: Record<string, unknown> = {};
      if (workflow.searchAttributes) {
        for (const [key, value] of Object.entries(
          workflow.searchAttributes
        )) {
          searchAttributes[key] = value;
        }
      }

      workflows.push({
        workflow_id: workflow.workflowId,
        run_id: workflow.runId,
        workflow_type: workflow.type,
        status: workflow.status.name,
        start_time: startTime,
        close_time: closeTime,
        duration_ms: durationMs,
        task_queue: workflow.taskQueue,
        search_attributes: searchAttributes,
      });

      count++;
      totalCount++;
    }
  } catch (error) {
    return handleError(error);
  }

  return {
    workflows,
    total_count: totalCount,
    has_more: hasMore,
  };
}

function handleError(error: unknown): never {
  const message = error instanceof Error ? error.message : String(error);

  if (message.includes("namespace") && message.includes("not found")) {
    throw new Error(
      `Namespace not found. Verify TEMPORAL_NAMESPACE is correct. Details: ${message}`
    );
  }

  if (message.includes("UNAVAILABLE") || message.includes("connect")) {
    throw new Error(
      `Cannot connect to Temporal server. Verify TEMPORAL_ADDRESS is correct and the server is running. Details: ${message}`
    );
  }

  throw new Error(`Failed to list workflows: ${message}`);
}

File: src/temporal_xray/tools/list_workflows.py
from __future__ import annotations

from ..temporal_client import get_temporal_client
from ..types import ListWorkflowsInput, ListWorkflowsOutput
from .shared import to_workflow_summary

STATUS_MAP = {
    "running": "Running",
    "completed": "Completed",
    "failed": "Failed",
    "timed_out": "TimedOut",
    "cancelled": "Canceled",
    "terminated": "Terminated",
}


def _escape_query_value(value: str) -> str:
    return value.replace("'", "''")


def _build_query(input: ListWorkflowsInput) -> str:
    if input.query:
        return input.query

    clauses: list[str] = []

    if input.workflow_type:
        clauses.append(f"WorkflowType = '{_escape_query_value(input.workflow_type)}'")

    if input.workflow_id:
        clauses.append(f"WorkflowId = '{_escape_query_value(input.workflow_id)}'")

    if input.status and input.status != "all":
        mapped = STATUS_MAP.get(input.status, input.status)
        clauses.append(f"ExecutionStatus = '{mapped}'")

    if input.start_time_from:
        clauses.append(f"StartTime >= '{input.start_time_from}'")

    if input.start_time_to:
        clauses.append(f"StartTime <= '{input.start_time_to}'")

    return " AND ".join(clauses)


async def list_workflows(input: ListWorkflowsInput) -> ListWorkflowsOutput:
    client = await get_temporal_client(input.namespace)
    query = _build_query(input)

    try:
        limit = input.limit or 10
        workflows = []
        has_more = False
        count = 0

        async for workflow in client.list_workflows(query=query or None):
            if count >= limit:
                has_more = True
                break
            workflows.append(to_workflow_summary(workflow))
            count += 1

        return ListWorkflowsOutput(
            workflows=workflows, total_count=count, has_more=has_more
        )
    except Exception as e:
        message = str(e)
        if "namespace" in message and "not found" in message:
            raise RuntimeError(
                f"Namespace not found. Verify TEMPORAL_NAMESPACE is correct. Details: {message}"
            ) from e
        if "UNAVAILABLE" in message or "connect" in message:
            raise RuntimeError(
                f"Cannot connect to Temporal server. Verify TEMPORAL_ADDRESS is correct and the server is running. Details: {message}"
            ) from e
        raise RuntimeError(f"Failed to list workflows: {message}") from e
